File: app/src/main/java/srjhlab/com/myownbarcode/MainActivity.java
package srjhlab.com.myownbarcode;

import android.Manifest;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Html;
import android.util.Log;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.google.zxing.integration.android.IntentIntegrator;
import com.google.zxing.integration.android.IntentResult;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;
import java.util.List;

import srjhlab.com.myownbarcode.Adapter.BarcodeRecyclerviewAdapter;
import srjhlab.com.myownbarcode.CommonUi.AddBarcodeInfoDialog;
import srjhlab.com.myownbarcode.CommonUi.AddFromImageDialog;
import srjhlab.com.myownbarcode.CommonUi.AddFromKeyDialog;
import srjhlab.com.myownbarcode.CommonUi.BarcodeFocusDialog;
import srjhlab.com.myownbarcode.CommonUi.BarcodeModifyDialog;
import srjhlab.com.myownbarcode.CommonUi.SelectDialog;
import srjhlab.com.myownbarcode.Item.BarcodeItem;
import srjhlab.com.myownbarcode.Item.SelectDialogItem;
import srjhlab.com.myownbarcode.Utils.CommonEventbusObejct;
import srjhlab.com.myownbarcode.Utils.CommonUtils;
import srjhlab.com.myownbarcode.Utils.ConstVariables;
import srjhlab.com.myownbarcode.Utils.DbHelper;

public class MainActivity extends AppCompatActivity implements BarcodeRecyclerviewAdapter.IOnClick {
    private final static String TAG = MainActivity.class.getSimpleName();

    private final int MY_PERMISSIONS_REQUEST = 1;

    private BarcodeItem mItem;
    private List<BarcodeItem> mItems;
    private RecyclerView mRecyclerview;
    private BarcodeRecyclerviewAdapter mAdapter;
    private DbHelper mDBHelper;
    private Bitmap mDefaultImage;
    private TextView mTextView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Log.d(TAG, "##### oncreate #####");

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        EventBus.getDefault().register(this);
        mDBHelper = new DbHelper(getApplicationContext(), "barcodeimg.db", null, 1);
        mRecyclerview = (RecyclerView) findViewById(R.id.recyclerView);
        mItems = new ArrayList<>();
        mAdapter = new BarcodeRecyclerviewAdapter(this);
        mRecyclerview.setHasFixedSize(true);
        mRecyclerview.setLayoutManager(new LinearLayoutManager(this));
        mRecyclerview.setAdapter(mAdapter);

        BitmapDrawable drawable = (BitmapDrawable) getResources().getDrawable(R.drawable.img_ref);
        mDefaultImage = drawable.getBitmap();

        mTextView = (TextView) findViewById(R.id.textview_license);
        String content = "오픈소스 라이센스";
        mTextView.setText(Html.fromHtml("<u>" + content + "</u>"));
        mTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, LicenseActivity.class);
                startActivity(intent);
                overridePendingTransition(R.anim.fade_in, R.anim.hold);
            }
        });
        initListItem();
    }

    @Override
    protected void onDestroy() {
        Log.d(TAG, "##### onDestroy #####");
        if (EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().unregister(this);
        }
        super.onDestroy();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        Log.d(TAG, "##### onActivityResult #####");
        super.onActivityResult(requestCode, resultCode, data);
        IntentResult result = IntentIntegrator.parseActivityResult(requestCode, resultCode, data);
        Log.d("TAG", "value : " + result.getContents() + "\n" + "format : " + result.getFormatName());

        if (result.getFormatName() == null) {
            return;
        }

        if (result.getFormatName().equals("CODE_39")
                || result.getFormatName().equals("CODE_93")
                || result.getFormatName().equals("CODE_128")
                || result.getFormatName().equals("EAN_8")
                || result.getFormatName().equals("EAN_13")
                || result.getFormatName().equals("PDF_417")
                || result.getFormatName().equals("UPC_A")
                || result.getFormatName().equals("UPC_E")
                || result.getFormatName().equals("CODABAR")
                || result.getFormatName().equals("ITF")
                || result.getFormatName().equals("QR_CODE")
                || result.getFormatName().equals("AZTEC")) {
            AddBarcodeInfoDialog.newInstance()
                    .setCommandType(AddBarcodeInfoDialog.MODE_ADD_BARCODE)
                    .setBarcodeItem(new BarcodeItem(result.getContents(), CommonUtils.convertBarcodeType(MainActivity.this, result.getFormatName())))
                    .show(getFragmentManager(), AddBarcodeInfoDialog.class.getSimpleName());

        } else {
            Toast.makeText(this, "지원하지 않는 포맷입니다", Toast.LENGTH_SHORT).show();
        }
    }

    void initListItem() {
        Log.d(TAG, "##### initListItem #####");
        if (mItems != null) {
            mItems.clear();
        }
        for (int i = 0; i <= mDBHelper.getLength(); i++) {
            if (mDBHelper.getBarcode()[i] != null) {
                mItem = new BarcodeItem(ConstVariables.ITEM_TYPE_BARCODE, mDBHelper.getId()[i], mDBHelper.getName()[i], mDBHelper.getColor()[i], mDBHelper.getValue()[i], makeBarcodeFromDB(mDBHelper.getBarcode()[i]));
            } else {
                mItem = new BarcodeItem(ConstVariables.ITEM_TYPE_EMPTY, 0, "새 바코드 추가", 0, " ", mDefaultImage);
            }
            mItems.add(mItem);
        }
        ((BarcodeRecyclerviewAdapter) mRecyclerview.getAdapter()).setItems(mItems);
    }

    void setBarcodeScan() {
        Log.d(TAG, "##### setBarcodeScan ######");

        int permissionCheck = ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA);

        if (permissionCheck == PackageManager.PERMISSION_DENIED) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
                if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.CAMERA)) {
                    ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA}, MY_PERMISSIONS_REQUEST);
                } else {
                    ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA}, MY_PERMISSIONS_REQUEST);
                }
            }

        } else {
            IntentIntegrator integrator = new IntentIntegrator(this);
            integrator.setOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
            integrator.addExtra("PROMPT_MESSAGE", "바코드를 사각형 안에 비춰주세요. \n"
                    + "스캔이 잘 안될 경우 주위 배경을 깔끔하게 해주세요. \n"
                    + "볼륨 UP/DOWN 버튼으로 플래시를 켜고 끌 수 있습니다.");
            integrator.setWide();
            integrator.initiateScan();
        }
    }

    public Bitmap makeBarcodeFromDB(byte[] b) {
        Log.d(TAG, "##### getBarcodeFromDb ######");

        Bitmap bitmap = BitmapFactory.decodeByteArray(b, 0, b.length);
        return bitmap;
    }

    @Override
    public void onItemClick(BarcodeItem item) {
        switch (item.getItemType()) {
            case ConstVariables.ITEM_TYPE_EMPTY:
                Log.d(TAG, "##### onItemClick ##### ITEM_TYPE_EMPTY");
                List<SelectDialogItem> mItems = new ArrayList<>();
                mItems.add(new SelectDialogItem(SelectDialogItem.INPUT_SELF));
                mItems.add(new SelectDialogItem(SelectDialogItem.INPUT_SCAN));
                mItems.add(new SelectDialogItem(SelectDialogItem.INPUT_IMAGE));

                if (mItems != null) {
                    final SelectDialog selectDialog = SelectDialog.newInstance();
                    selectDialog.setItems(mItems);
                    selectDialog.show(getFragmentManager(), selectDialog.getClass().getSimpleName());
                }
                break;
            case ConstVariables.ITEM_TYPE_BARCODE:
                Log.d(TAG, "##### onItemClick ##### ITEM_TYPE_BARCODE");
                if (item != null) {
                    final BarcodeFocusDialog barcodeFocusDiaolg = BarcodeFocusDialog.newInstance();
                    barcodeFocusDiaolg
                            .setItem(item)
                            .setType(barcodeFocusDiaolg.VIEW_TYPE_FOCUS)
                            .show(getFragmentManager(), barcodeFocusDiaolg.getClass().getSimpleName());
                }

                break;
        }
    }

    @Override
    public void onItemLongClick(BarcodeItem item) {
        switch (item.getItemType()) {
            case ConstVariables.ITEM_TYPE_EMPTY:
                Log.d(TAG, "##### onItemLongClick ##### ITEM_TYPE_EMPTY");
                break;
            case ConstVariables.ITEM_TYPE_BARCODE:
                Log.d(TAG, "##### onItemLongClick ##### ITEM_TYPE_BARCODE");
                if (item != null) {
                    final BarcodeModifyDialog barcodeModifyDialog = BarcodeModifyDialog.newInstance();
                    barcodeModifyDialog.setItem(item);
                    barcodeModifyDialog.show(getFragmentManager(), barcodeModifyDialog.getClass().getSimpleName());
                }
                break;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        switch (requestCode) {
            case MY_PERMISSIONS_REQUEST:
                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    IntentIntegrator integrator = new IntentIntegrator(this);
                    integrator.setOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
                    integrator.addExtra("PROMPT_MESSAGE", "바코드를 사각형 안에 비춰주세요. \n" + "스캔이 잘 안될 경우 주위 배경을 깔끔하게 해주세요.");
                    integrator.initiateScan();
                } else {
                    Toast.makeText(this, "카메라 사용 권한을 허용해야 카메라 스캔을 사용할 수 있습니다.", Toast.LENGTH_SHORT).show();
                }
                return;
        }
    }

    /*
     * START EVENTBUS RECEIVE
     * */

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(CommonEventbusObejct object) {
        switch (object.getType()) {
            case ConstVariables.EVENTBUS_ADD_NEW_BARCODE:
                Log.d(TAG, "##### EVENTBUS_ADD_NEW_BARCODE #####");
                /*List<SelectDialogItem> mItems = new ArrayList<>();
                mItems.add(new SelectDialogItem(SelectDialogItem.INPUT_SELF));
                mItems.add(new SelectDialogItem(SelectDialogItem.INPUT_SCAN));
                mItems.add(new SelectDialogItem(SelectDialogItem.INPUT_IMAGE));

                if (mItems != null) {
                    final SelectDialog selectDialog = SelectDialog.newInstance();
                    selectDialog.setItems(mItems);
                    selectDialog.show(getFragmentManager(), selectDialog.getClass().getSimpleName());
                }*/
                break;
            case ConstVariables.EVENTBUS_SHOW_BARCODE:
                Log.d(TAG, "##### EVENTBUS_SHOW_BARCODE #####");
                break;
            case ConstVariables.EVENTBUS_MODIFY_BARCODE:
                Log.d(TAG, "##### EVENTBUS_MODIFY_BARCODE #####");
                break;
            case ConstVariables.EVENTBUS_ADD_FROM_KEY:
                Log.d(TAG, "##### EVENTBUS_ADD_FROM_KEY #####");
                AddFromKeyDialog dialog = AddFromKeyDialog.newInstance();
                dialog.show(getFragmentManager(), dialog.getClass().getSimpleName());
                break;
            case ConstVariables.EVENTBUS_ADD_FROM_CCAN:
                Log.d(TAG, "##### EVENTBUS_ADD_FROM_CCAN #####");
                setBarcodeScan();
                break;
            case ConstVariables.EVENTBUS_ADD_FROM_IMAGE:
                Log.d(TAG, "##### EVENTBUS_ADD_FROM_IMAGE #####");
                AddFromImageDialog addFromImageDialog = AddFromImageDialog.newInstance();
                addFromImageDialog.show(getFragmentManager(), addFromImageDialog.getClass().getSimpleName());
                break;
            case ConstVariables.EVENTBUS_ADD_BARCODE:
                Log.d(TAG, "##### EVENTBUS_ADD_BARCODE #####");
                AddBarcodeInfoDialog.newInstance()
                        .setBarcodeItem((BarcodeItem) object.getVal())
                        .setCommandType(AddBarcodeInfoDialog.MODE_ADD_BARCODE)
                        .show(getFragmentManager(), AddBarcodeInfoDialog.class.getSimpleName());
                break;
            case ConstVariables.EVENTBUS_ADD_BARCODE_SUCCESS:
                initListItem();
                Log.d(TAG, "##### EVENTBUS_ADD_BARCODE_SUCCESS #####");
                break;
            case ConstVariables.EVENTBUS_EDIT_BARCODE:
                Log.d(TAG, "##### EVENTBUS_EDIT_BARCODE #####");
                AddBarcodeInfoDialog.newInstance()
                        .setBarcodeItem((BarcodeItem) object.getVal())
                        .setCommandType(AddBarcodeInfoDialog.MODE_EDIT_BARCODE)
                        .show(getFragmentManager(), AddBarcodeInfoDialog.class.getSimpleName());
                break;
            case ConstVariables.EVENTBUS_DELETE_BARCODE:
                Log.d(TAG, "##### EVENTBUS_DELETE_BARCODE #####");
                mDBHelper.delete(((BarcodeItem)object.getVal()).getBarcodeId());
                initListItem();
                break;
            case ConstVariables.EVENTBUS_SHARE_BARCODE:
                Log.d(TAG, "##### EVENTBUS_SHARE_BARCODE #####");

                final BarcodeFocusDialog barcodeFocusDiaolg = BarcodeFocusDialog.newInstance();
                barcodeFocusDiaolg
                        .setItem(((BarcodeItem)object.getVal()))
                        .setType(barcodeFocusDiaolg.VIEW_TYPE_SHARE)
                        .show(getFragmentManager(), barcodeFocusDiaolg.getClass().getSimpleName());
        }
    }
    /*
     * END EVENTBUS RECEIVE
     * */
}

File: app/src/main/kotlin/Module/Main/MainActivity.kt
package srjhlab.com.myownbarcode

import Model.ActivityResultEvent
import Module.Main.Main
import Module.Main.MainPresenter
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.Gravity
import android.view.View
import com.google.firebase.auth.FirebaseAuth
import com.google.zxing.integration.android.IntentIntegrator
import kotlinx.android.synthetic.main.activity_main.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.jetbrains.anko.selector
import org.jetbrains.anko.toast
import srjhlab.com.myownbarcode.Dialog.AddBarcodeInfoDialog
import srjhlab.com.myownbarcode.Dialog.AddFromImageDialog
import srjhlab.com.myownbarcode.Dialog.AddFromKeyDialog
import srjhlab.com.myownbarcode.Dialog.FocusDialog
import srjhlab.com.myownbarcode.Item.BarcodeItem
import srjhlab.com.myownbarcode.Module.MyBarcode.MyBarcodeFragment
import srjhlab.com.myownbarcode.Module.Settings.SettingsFragment
import srjhlab.com.myownbarcode.Utils.CommonEventbusObejct
import srjhlab.com.myownbarcode.Utils.CommonUtils
import srjhlab.com.myownbarcode.Utils.ConstVariables

class MainActivity : AppCompatActivity(), View.OnClickListener, Main.view {
    val TAG = this.javaClass.simpleName
    private lateinit var mAuth: FirebaseAuth
    private val PERMISSIONS_REQUEST = 1
    private val FRAGMENT_STATE_BARCODE_LIST = 1
    private val FRAGMENT_STATE_SETTINGS = 2
    private lateinit var mPresenter: MainPresenter
    private lateinit var mMyBarcodeFragment: MyBarcodeFragment
    private lateinit var mSettingsFragment: SettingsFragment
    private lateinit var mContentView: View

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        mAuth = FirebaseAuth.getInstance()
        EventBus.getDefault().register(this)
        mPresenter = MainPresenter(this)
        mMyBarcodeFragment = MyBarcodeFragment()
        mSettingsFragment = SettingsFragment()
        initializeUI()
        hanleFragment(FRAGMENT_STATE_BARCODE_LIST)
    }

    private fun initializeUI() {
        Log.d(TAG, "##### initializeUI #####")
        mContentView = layout_content as View
        img_drawer_icon.setOnClickListener(this)
        textview_drawer_home.setOnClickListener(this)
        textview_drawer_settings.setOnClickListener(this)
        textview_drawer_license.setOnClickListener(this)
    }

    override fun onStart() {
        Log.d(TAG, "##### onStart #####")
        super.onStart()
        val currentUser = mAuth.currentUser
        if (currentUser != null) {
            Log.d(TAG, "##### onStart currentUserId : ${currentUser.email}#####")
        }
    }

    override fun onDestroy() {
        Log.d(TAG, "##### onDestroy ####")
        super.onDestroy()
        if (EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().unregister(this)
        }
    }

    override fun onClick(v: View) {
        Log.d(TAG, "##### onClick #####")
        val viewId = v.id

        when (viewId) {
            img_drawer_icon.id -> layout_drawer.openDrawer(Gravity.LEFT)
            textview_drawer_home.id -> {
                hanleFragment(FRAGMENT_STATE_BARCODE_LIST)
                layout_drawer.closeDrawer(Gravity.LEFT)
            }
            textview_drawer_settings.id -> {
                hanleFragment(FRAGMENT_STATE_SETTINGS)
                layout_drawer.closeDrawer(Gravity.LEFT)
            }
            textview_drawer_license.id -> Log.d(TAG, "##### onItemCLick #####")
            else -> return
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        Log.d(TAG, "##### onRequestPermissionRequest #####")
        when (requestCode) {
            PERMISSIONS_REQUEST -> {
                if (grantResults.isNotEmpty() && grantResults[0].equals(PackageManager.PERMISSION_GRANTED)) {
                    mPresenter.requestBarcodeScan()
                } else {
                    toast(R.string.string_require_camera_permission)
                }
            }
        }
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        Log.d(TAG, "##### onActivityResult #####")
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == ConstVariables.RC_SIGN_IN) {
            EventBus.getDefault().post(CommonEventbusObejct(ConstVariables.EVENTBUS_ON_ACTIBITY_RESULT, ActivityResultEvent(requestCode, resultCode, data)))
            return
        }

        val result = IntentIntegrator.parseActivityResult(requestCode, resultCode, data)
        Log.d(TAG, "##### onActivityResult ##### value : " + result.contents + " format : " + result.formatName)
        when (result.formatName) {
            null -> return
            "CODE_39", "CODE_93", "CODE_128", "EAN_8", "EAN_13", "PDF_417", "UPC_A", "UPC_E", "CODABAR", "ITF", "QR_CODE", "AZTEC" ->
                AddBarcodeInfoDialog()
                        .setCommandType(AddBarcodeInfoDialog.MODE_ADD_BARCODE)
                        .setBarcodeItem(BarcodeItem(result.contents, CommonUtils.convertBarcodeType(this, result.formatName).toLong()))
                        .show(fragmentManager, this.javaClass.simpleName)
            else -> toast(R.string.string_not_supported_format)
        }
    }

    override fun onResultBackPressed(result: Boolean, msg: Int) {
        Log.d(TAG, "##### onResultBackPressed #####")
        if (result) {
            this.finish()
        } else {
            toast(resources.getString(msg))
        }
    }

    private fun hanleFragment(state: Int) {
        Log.d(TAG, "##### hanleFragment ##### state : $state")
        val fm = supportFragmentManager
        val ft = fm.beginTransaction()
        var fr = Fragment()
        ft.setCustomAnimations(android.R.animator.fade_in, android.R.animator.fade_out)

        when (state) {
            FRAGMENT_STATE_BARCODE_LIST -> {
                fr = MyBarcodeFragment()
            }

            FRAGMENT_STATE_SETTINGS -> {
                fr = SettingsFragment()
            }
        }
        fm.popBackStack()
        if (fr.isAdded) {
            ft.show(fr)
        } else {
            ft.add(mContentView.id, fr, fr.javaClass.simpleName)
        }
        ft.addToBackStack(null)


        ft.commit()
    }

    override fun onBackPressed() {
        Log.d(TAG, "##### onBackPressed #####")
        mPresenter.requestBackPressed()
    }

    @Subscribe
    fun onEVent(busObject: CommonEventbusObejct) {
        when (busObject.type) {
            ConstVariables.EVENTBUS_ADD_FROM_SCAN -> {
                Log.d(TAG, "##### EVENTBUS_ADD_FROM_SCAN #####")
                mPresenter.requestBarcodeScan()
            }
            ConstVariables.EVENTBUS_ADD_FROM_KEY -> {
                Log.d(TAG, "##### EVENTBUS_ADD_FROM_KEY #####")
                AddFromKeyDialog()
                        .show(fragmentManager, this.javaClass.simpleName)
            }
            ConstVariables.EVENTBUS_ADD_FROM_IMAGE -> {
                Log.d(TAG, "##### EVENTBUS_ADD_FROM_IMAGE #####")
                AddFromImageDialog()
                        .show(fragmentManager, this.javaClass.simpleName)
            }
            ConstVariables.EVENTBUS_ADD_BARCODE -> {
                Log.d(TAG, "##### EVENTBUS_ADD_BARCODE #####")
                AddBarcodeInfoDialog()
                        .setBarcodeItem(busObject.`val` as BarcodeItem)
                        .setCommandType(AddBarcodeInfoDialog.MODE_ADD_BARCODE)
                        .show(fragmentManager, this.javaClass.simpleName)
            }
            ConstVariables.EVENTBUS_EDIT_BARCODE -> {
                Log.d(TAG, "##### EVENTBUS_EDIT_BARCODE #####")
                AddBarcodeInfoDialog()
                        .setBarcodeItem(busObject.`val` as BarcodeItem)
                        .setCommandType(AddBarcodeInfoDialog.MODE_EDIT_BARCODE)
                        .show(fragmentManager, this.javaClass.simpleName)
            }
            ConstVariables.EVENTBUS_SHARE_BARCODE -> {
                Log.d(TAG, "##### EVENTBUS_SHARE_BARCODE #####")
                FocusDialog()
                        .setBarcodeItem(busObject.`val` as BarcodeItem)
                        .setCommandType(FocusDialog.VIEW_TYPE_SHARE)
                        .show(fragmentManager, this.javaClass.simpleName)
            }
            ConstVariables.EVENTBUS_CLICK_BARCODELIST -> {
                Log.d(TAG, "##### EVENTBUS_CLICK_BARCODELIST #####")
                if (layout_drawer.isDrawerOpen(Gravity.LEFT)) {
                    return
                }
                val item = busObject.`val` as BarcodeItem
                when (item.itemType) {
                    ConstVariables.ITEM_TYPE_EMPTY -> {
                        val menuList = listOf(getString(R.string.string_input_key), getString(R.string.string_input_scan), getString(R.string.string_input_image))
                        selector("", menuList) { dialogInterface, i ->
                            when (i) {
                                0 -> {
                                    AddFromKeyDialog().show(fragmentManager, this.javaClass.simpleName)
                                }
                                1 -> {
                                    mPresenter.requestBarcodeScan()
                                }
                                2 -> {
                                    AddFromImageDialog().show(fragmentManager, this.javaClass.simpleName)
                                }
                            }
                        }
                    }
                    ConstVariables.ITEM_TYPE_BARCODE -> {
                        FocusDialog().setBarcodeItem(item).setCommandType(FocusDialog.VIEW_TYPE_FOCUS).show(fragmentManager, this.javaClass.simpleName)
                    }
                }
            }
            ConstVariables.EVENTBUS_LONGCLICK_BARCODELIST -> {
                Log.d(TAG, "##### EVENTBUS_LONGCLICK_BARCODELIST #####")
                val item = busObject.`val` as BarcodeItem
                when (item.itemType) {
                    ConstVariables.ITEM_TYPE_EMPTY -> Log.d(TAG, "##### onItemLongClick ##### ITEM_TYPE_EMPTY")
                    ConstVariables.ITEM_TYPE_BARCODE -> {
                        Log.d(TAG, "##### onItemLongClick #####")
                        val menuList = listOf(getString(R.string.string_modify_barcode), getString(R.string.string_delete_barcode), getString(R.string.string_share_barcode))
                        selector("", menuList) { dialogInterface, i ->
                            when (i) {
                                0 -> {
                                    AddBarcodeInfoDialog()
                                            .setBarcodeItem(busObject.`val` as BarcodeItem)
                                            .setCommandType(AddBarcodeInfoDialog.MODE_EDIT_BARCODE)
                                            .show(fragmentManager, this.javaClass.simpleName)
                                }
                                1 -> {
                                    EventBus.getDefault().post(CommonEventbusObejct(ConstVariables.EVENTBUS_DELETE_BARCODE, busObject.`val` as BarcodeItem))
                                }
                                2 -> {
                                    FocusDialog()
                                            .setBarcodeItem(busObject.`val` as BarcodeItem)
                                            .setCommandType(FocusDialog.VIEW_TYPE_SHARE)
                                            .show(fragmentManager, this.javaClass.simpleName)
                                }
                            }
                        }
                    }
                }

            }
        }
    }
}